lib/filewatch/observing_base.rb

```diff
@@ -0,0 +1,101 @@
+# encoding: utf-8
+
+## Interface API topology
+# ObservingBase module (this file)
+#   is a module mixin proving common constructor and external API for File Input Plugin interaction
+#   calls build_specific_processor on ObservingRead or ObservingTail
+# ObservingRead and ObservingTail
+#   provides the External API method subscribe(observer = NullObserver.new)
+#   build_specific_processor(settings) - provide a Tail or Read specific Processor.
+# TailMode::Processor or ReadMode::Processor
+#   initialize_handlers(sincedb_collection, observer) - called when the observer subscribes to changes in a Mode,
+#      builds mode specific handler instances with references to the observer
+#   process_closed(watched_files) - provide specific processing of watched_files in the closed state
+#   process_ignored(watched_files) - provide specific processing of watched_files in the ignored state
+#   process_watched(watched_files) - provide specific processing of watched_files in the watched state
+#   process_active(watched_files)  - provide specific processing of watched_files in the active state
+#      These methods can call "handler" methods that delegate to the specific Handler classes.
+# TailMode::Handlers module namespace
+#   contains the Handler classes that deals with Tail mode file lifecycle "events".
+#   The TailMode::Handlers::Base
+#      handle(watched_file) - this method calls handle_specifically defined in a subclass
+#      handle_specifically(watched_file) - this is a noop method
+#      update_existing_specifically(watched_file, sincedb_value) - this is a noop method
+#   Each handler extends the Base class to provide specific implementations of these two methods:
+#      handle_specifically(watched_file)
+#      update_existing_specifically(watched_file, sincedb_value)
+# ReadMode::Handlers module namespace
+#   contains the Handler classes that deals with Read mode file lifecycle "events".
+#   The ReadMode::Handlers::Base
+#      handle(watched_file) - this method calls handle_specifically defined in a subclass
+#      handle_specifically(watched_file) - this is a noop method
+#   Each handler extends the Base class to provide specific implementations of this method:
+#      handle_specifically(watched_file)
+
+module FileWatch
+  module ObservingBase
+    attr_reader :watch, :sincedb_collection, :settings
+
+    def initialize(opts={})
+      options = {
+        :sincedb_write_interval => 10,
+        :stat_interval => 1,
+        :discover_interval => 5,
+        :exclude => [],
+        :start_new_files_at => :end,
+        :delimiter => "\n",
+        :file_chunk_count => FIXNUM_MAX,
+        :file_sort_by => "last_modified",
+        :file_sort_direction => "asc",
+      }.merge(opts)
+      unless options.include?(:sincedb_path)
+        options[:sincedb_path] = File.join(ENV["HOME"], ".sincedb") if ENV.include?("HOME")
```

Comment:
do we use the `HOME` env var? isn't it `LOGSTASH_HOME` and `LS_HOME` per https://github.com/elastic/logstash/blob/57e7a8a56bdbf81647d6e116e75263796d6daf47/bin/logstash.lib.sh#L44-L48